Fixup GDI object leaks

mstv opened the pull request:
Fixes #9166

## Proposed changes

Fixup GDI object leaks:
- Release DC retrieved using `GetWindowDC`
- Dispose resources in `WindowsJumpListManager.MakeIcon`
- Correctly convert `Bitmap` to `Icon`
- Cache `Icons` drawn for Taskbar overlay / Jump List

## Test methodology <!-- How did you ensure quality? -->

- manual

## Test environment(s) <!-- Remove any that don't apply -->

- Git Extensions 33.33.33
- Build 99649dc430eb85fbacda92046a639af8d1c552dc
- Git 2.31.1.windows.1
- Microsoft Windows NT 10.0.19042.0
- .NET Framework 4.8.4360.0
- DPI 96dpi (no scaling)

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

RussKie commented on GitUI/Interops/User32/GetWindowDC.cs:
```diff
@@ -7,5 +7,8 @@ internal static partial class NativeMethods
         public static extern IntPtr GetWindowDC(IntPtr hwnd);
+
+        [DllImport(Libraries.User32)]
+        public static extern void ReleaseDC(IntPtr hwnd, IntPtr dc);
```
Move it to GitUI/Interops/User32/ReleaseDC.cs

mstv replied:
Yes, of course.

RussKie reviewed (approved):
🚀

mstv commented:
I have rebased and added a proposal https://github.com/gitextensions/gitextensions/commit/d81ce7544cf293ca4093177c2c1e3094a3ff054a with minor changes to "Correctly convert bitmap to icon":
- Rename extension method to just `ToBitmap` (like `ToString`)
- Replace `var`

gerhardol reviewed (commented):
+1
An alternative approach could have been to not recreate the icon if it is the same as before. This has less leaks, just slightly higher memory use though.

mstv commented:
> An alternative approach could have been to not recreate the icon if it is the same as before.

That has already been implemented: refer to `LastBrush`. But is has only very little effect because the status is invalidated often.

> This has less leaks, just slightly higher memory use though.

It is limited to a low number of icons since we have a defined set of both overlay colors and images for the Jump List buttons.

sharwell commented on GitUI/BitmapExtensions.cs:
```diff
@@ -0,0 +1,34 @@
+{
+    public static class BitmapExtensions
+    {
+        public static Icon BitmapToIcon(this Bitmap bitmap)
```
💡 As an extension method, `Bitmap` seems redundant.

```suggestion
        public static Icon ToIcon(this Bitmap bitmap)
```

mstv replied:
I forgot to push...

sharwell commented on GitUI/CommandsDialogs/FormBrowse.cs:
```diff
@@ -104,6 +104,8 @@ public sealed partial class FormBrowse : GitModuleForm, IBrowseRepo
 
         [CanBeNull] private TabPage _consoleTabPage;
 
+        private Dictionary<Brush, Icon> _overlayIconByBrush = new();
```
```suggestion
        private readonly Dictionary<Brush, Icon> _overlayIconByBrush = new();
```

mstv replied:
Done. (I find C++ `const` clearer than `readonly`.)

sharwell commented on GitUI/WindowsJumpListManager.cs:
```diff
@@ -24,6 +25,7 @@ public sealed class WindowsJumpListManager : IDisposable
         private ThumbnailToolBarButton _pushButton;
         private ThumbnailToolBarButton _pullButton;
         private string _deferredAddToRecent;
+        private static Dictionary<Image, Icon> _iconByImage = new();
```
```suggestion
        private static readonly Dictionary<Image, Icon> _iconByImage = new();
```

mstv replied:
Done

sharwell commented on GitUI/UserControls/TextBoxEx.cs:
```diff
@@ -60,24 +60,22 @@ protected override void WndProc(ref Message m)
             {
                 case NativeMethods.WM_NCPAINT:
                 case NativeMethods.WM_PAINT:
```
❓ Can we not just set `UserPaint` style and move the draw code to `OnPaint`?

mstv replied:
This should be done in a dedicated PR on the main dev branch.

RussKie commented:
Champion 💯
